### DISCUSSION TITLE
SAHI Tiled Inference - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#4769] [Docs]
# SAHI Tiled Inference - Ultralytics YOLOv8 Docs

A comprehensive guide on how to use YOLOv8 with SAHI for standard and sliced inference in object detection tasks.

https://docs.ultralytics.com/guides/sahi-tiled-inference/

**Comment by GarethWright** [comment]
Will this be available via the cli?

**Reply by RizwanMunawar** [reply]
@GarethWright SAHI support is not available with CLI, We may release CLI for this in future.

Thanks
Ultralytics Team!

**Comment by ExtReMLapin** [comment]
Training with SAHI would be top notch.

**Reply by pderrenger** [reply]
@ExtReMLapin absolutely! Training with SAHI can significantly enhance the object detection capabilities of YOLOv8, especially when dealing with large-scale or high-resolution images. It's a powerful combination for tackling complex detection tasks. 🚀 If you need any guidance on setting it up or optimizing your workflow, feel free to check out the [docs](https://docs.ultralytics.com/guides/sahi-tiled-inference/) or ask here! Happy training! 😊👨‍💻

**Comment by jehan88** [comment]
how can i use SAHI with my weight,   best.pt?

**Reply by fcakyon** [reply]
You can use it this way:

```
yolov8_model_path = 'best.pt'

detection_model = AutoDetectionModel.from_pretrained(
    model_type='yolov8',
    model_path=yolov8_model_path,
    confidence_threshold=0.3,
    device="cpu",  # or 'cuda:0'
)
```

**Reply by dimaspratama08** [reply]
Hi sir, i have trained a model and it works decently when i tried to inference without integrating SAHI at all. But when i tried to integrate it using sahi, which is this code:
# With an image path
result = get_prediction("demo_data/small-vehicles1.jpeg", detection_model)

# With a numpy image
result = get_prediction(read_image("demo_data/small-vehicles1.jpeg"), detection_model)

the output didnt show any prediction at all

**Reply by pderrenger** [reply]
@dimaspratama08 hello! If you're not seeing any predictions after integrating SAHI, it could be due to a few reasons:

1. **Slice Size**: If the slice dimensions are too small, they may not contain the objects you want to detect. Ensure that the slice size is large enough to capture the objects.

2. **Overlap Ratios**: If the overlap ratios are too small, objects on the edges of slices might not be detected. Increasing the overlap can help capture objects that are split between slices.

3. **Confidence Threshold**: The confidence threshold might be set too high, causing detections with lower confidence to be discarded. Try lowering the threshold.

4. **Model Compatibility**: Ensure that the `best.pt` model is compatible with SAHI and that the model is correctly loaded.

5. **Image Path**: Verify that the image path is correct and the image is accessible.

6. **Model Device**: Ensure that the model is loaded onto the correct device (CPU/GPU) that you're intending to use for inference.

Here's a modified example with sliced prediction that includes the slice size and overlap ratios:

```python
from sahi.predict import get_sliced_prediction

result = get_sliced_prediction(
    image_path="demo_data/small-vehicles1.jpeg",
    detection_model=detection_model,
    slice_height=1024,  # adjust based on your image size and objects
    slice_width=1024,   # adjust based on your image size and objects
    overlap_height_ratio=0.2,  # adjust as needed
    overlap_width_ratio=0.2    # adjust as needed
)
```

Make sure to adjust the `slice_height`, `slice_width`, `overlap_height_ratio`, and `overlap_width_ratio` according to your specific use case.

If you're still facing issues, please check the SAHI and YOLOv8 documentation for more detailed instructions, or consider reaching out to the SAHI community for support.

**Comment by jehanjen** [comment]
how i can save txt for detetcion object from sahi with yolov8 ?

**Reply by jehanjen** [reply]
thanks for your answer,
how can i save crop image after detection.

**Reply by glenn-jocher** [reply]
@jehan88 see the `result.save_crop()` method at https://docs.ultralytics.com/reference/engine/results/?h=save_crop#ultralytics.engine.results.Results.save_crop

**Comment by cnydo** [comment]
It would be great if sahi is integrated into training with yolov8

**Reply by rusvagzur** [reply]
I think you can just use the native SAHI source code to create a dataset with slices and use YOLOv8 as with any other dataset

**Reply by fcakyon** [reply]
I am the author of SAHI. What @rusvagzur mentions is a great best practice, you can try it @cnydo.

**Reply by IrDIE** [reply]
I implemented validation with SAHI for YOLO8 
https://github.com/IrDIE/YOLO8_SAHI/tree/master

:blue_heart: If you like it - press :star:

**Reply by pderrenger** [reply]
That's fantastic to hear about your initiative, @IrDIE! It's always inspiring to see community members contributing and building upon the YOLOv8 framework. While I can't directly endorse or review external projects, I encourage you to keep collaborating and sharing your work with the community. Your efforts help everyone grow and improve. Keep up the great work! 💙🌟

**Reply by cnydo** [reply]
very nice

**Reply by glenn-jocher** [reply]
Thanks for the kind words! 😊 If you're looking to dive deeper into using YOLOv8 with SAHI for sliced inference, our [Ultralytics documentation](https://docs.ultralytics.com/guides/sahi-tiled-inference/) is definitely the place to check out. It's packed with useful info to help you get started, including setup instructions, code snippets, and explanations on how SAHI can enhance YOLOv8's capabilities for large-scale imagery. Happy detecting! 🚀

**Comment by rusvagzur** [comment]
does this allow batch>1 inference? I mean, batch inference in the GPU all slices at the same time to get also an inference speed advantage compared to doing inference on a huge image without SAHI

**Reply by fcakyon** [reply]
SAHI currently does not support batch>1 inference.

**Comment by stefp** [comment]
Hi, it seems  that is only applicable to object detection models.

I would like to apply to an image classification model. Are there any solutions for that?

Thanks

**Reply by andyrey** [reply]
I have many annotated small objects in my dataset.  I want to use SAHI during inference. Should I employ some SAHI-specific methods during training?

**Reply by fcakyon** [reply]
@stefp there is no solution for that at the moment. But stay tuned for upcoming updates!

**Reply by fcakyon** [reply]
Hey @andyrey, I am the author of SAHI. You can try random crop augmentations during training to increase the SAHI inference performance.

**Reply by andyrey** [reply]
Hello @fcakyon, Is there "random crop augmentations" in yolo 8 yaml file? I didn't found..

**Reply by ExtReMLapin** [reply]
@andyrey with albumentations yes

**Comment by mrcrzt** [comment]
Great work! is there a way to pass a labels array for prediction? to exclude all the other unwanted labels?

**Reply by glenn-jocher** [reply]
@mrcrzt yes, you can filter predictions by specific class labels using the `classes` argument during prediction. This argument accepts a list of class indices that you want to include in the predictions. Here's how you can use it:

In Python:
```python
from ultralytics import YOLO

 # Load a pretrained YOLOv8 model
 model = YOLO('yolov8n.pt')

  # Define the class indices you want to include in the predictions
  # For example, if you only want to include classes 0 and 2, you would pass [0, 2]
  desired_classes = [0, 2]

  # Run predictions while filtering for the desired classes
  results = model.predict('path/to/image.jpg', classes=desired_classes)
  ```

With CLI:
```bash
# Predict with an official YOLOv8 model while filtering for classes 0 and 2
yolo detect predict model=yolov8n.pt source='path/to/image.jpg' classes=0,2
```

Remember to replace `desired_classes` with the actual indices of the classes you want to include from your dataset's class labels. 🎯

**Comment by riobastian09** [comment]
can it returned the segmentation (masks) coordinate? not only bbox?

**Reply by glenn-jocher** [reply]
@riobastian09 yes, YOLOv8 can return segmentation masks in addition to bounding boxes. When using a YOLOv8 model that supports segmentation, such as YOLOv8n-seg, the model can predict both bounding boxes and segmentation masks for detected objects.

The segmentation masks are provided as binary masks that indicate the pixels belonging to each detected object. These masks can be accessed through the `masks` attribute of the `Results` object returned by the model's prediction method.

Here's a concise example of how to access segmentation masks:

```python
from ultralytics import YOLO

# Load a pretrained segmentation model
model = YOLO('yolov8n-seg.pt')

# Perform inference on an image
results = model('path/to/image.jpg')

# Access segmentation masks
for result in results:
    masks = result.masks.data  # Binary masks for each detected object
    segments = result.masks.xy  # Segments (pixels) for each mask
```

Each mask in the `masks` attribute is a binary mask corresponding to a detected object, and `segments` provides the pixel coordinates for the segmented areas. You can use these masks for further processing or visualization as needed.

**Comment by IrDIE** [comment]
Hi !
@rusvagzur , I have a repo for both SAHI inference and SAHI validation.
https://github.com/IrDIE/YOLO8_SAHI
it use gpu, you can try it 
If you like it press  :star:
Any recomendations for improvements are welcomed  :purple_heart:

**Reply by fcakyon** [reply]
Amazing job! 🚀

**Reply by zoldaten** [reply]
@IrDIE https://github.com/IrDIE/YOLO8_SAHI/issues/5 - cant change labels could you help ?

**Comment by zoldaten** [comment]
hi.
i try SAHI with yolov8m-oiv7.pt.
Model has 600 classes. So i put them in dict like coco (i provide here only format, i know that 600 names is much bigger):
`{'0': 'man', '1': 'bicycle', '2': 'car', '3': 'motorcycle', '4': 'airplane', '5': 'bus', '6': 'train', '7': 'truck', '8': 'boat', '9': 'traffic light', '10': 'fire hydrant', '11': 'stop sign', '12': 'parking meter', '13': 'bench', '14': 'bird', '15': 'cat', '16': 'dog', '17': 'horse', '18': 'sheep', '19': 'cow', '20': 'elephant', '21': 'bear', '22': 'zebra', '23': 'giraffe', '24': 'backpack', '25': 'umbrella', '26': 'handbag', '27': 'tie', '28': 'suitcase', '29': 'frisbee', '30': 'skis', '31': 'snowboard', '32': 'sports ball', '33': 'kite', '34': 'baseball bat', '35': 'baseball glove', '36': 'skateboard', '37': 'surfboard', '38': 'tennis racket', '39': 'bottle', '40': 'wine glass', '41': 'cup', '42': 'fork', '43': 'knife', '44': 'spoon', '45': 'bowl', '46': 'banana', '47': 'apple', '48': 'sandwich', '49': 'orange', '50': 'broccoli', '51': 'carrot', '52': 'hot dog', '53': 'pizza', '54': 'donut', '55': 'cake', '56': 'chair', '57': 'couch', '58': 'potted plant', '59': 'bed', '60': 'dining table', '61': 'toilet', '62': 'tv', '63': 'laptop', '64': 'mouse', '65': 'remote', '66': 'keyboard', '67': 'cell phone', '68': 'microwave', '69': 'oven', '70': 'toaster', '71': 'sink', '72': 'refrigerator', '73': 'book', '74': 'clock', '75': 'vase', '76': 'scissors', '77': 'teddy bear', '78': 'hair drier', '79': 'toothbrush'}`

set to detection model:
```
detection_model = AutoDetectionModel.from_pretrained(
    model_type='yolov8',    
    confidence_threshold=0.1,    
    model_path='models/yolov8m-oiv7.pt',     
    category_mapping=category_mapping,
    device="cpu", # or 'cuda:0'
)
```

but in output i see again coco classes.
`result = get_prediction("686.jpg", detection_model)`

i also tried put to detection_model .yaml - ` class_names_yaml_path="tests/data/coco_utils/coco_class_names.yaml"`
but code crashes.

how to change classes for yolov8m-oiv7.pt ?

**Reply by fcakyon** [reply]
Hey @zoldaten, creator of `sahi` here 👍🏻 

You should use [`category_remapping`](https://github.com/obss/sahi/blob/1240bab3b838fd30929474debe9e086f15bd07af/sahi/models/base.py#L43) instead of `category_mapping` if you want to remap/filter the original categories of `yolov8m-oiv7` model.

From the docstr: `Remap category ids based on category names, after performing inference e.g. {"car": 3}`

**Reply by zoldaten** [reply]
not actually what i asked but i found the way:
```
category_mapping=dict()
with open ('600_eng.txt', encoding='utf-8') as f:
    for count, l in enumerate (f.readlines()):
        category_mapping[str(count)]=l.rstrip()[5:].strip()
###https://docs.ultralytics.com/datasets/detect/open-images-v7/#dataset-yaml
```
 
and 
```
detection_model = AutoDetectionModel.from_pretrained( model_type="yolov8", \
                                                          model_path='yolov8m-oiv7.pt', \
                                                          confidence_threshold=0.2, \
                                                          category_mapping=category_mapping,
                                                          device="gpu")
```

and output:
![out — копия](https://github.com/user-attachments/assets/1cdeef3c-bff9-4d84-ac5c-346382700b6d)


dont know why bell pepper, but detection ok enough. thanks!

**Reply by glenn-jocher** [reply]
Great to hear you found a solution! For the "bell pepper" label, ensure your `600_eng.txt` matches the [Open Images V7 class list](https://docs.ultralytics.com/datasets/detect/open-images-v7/) exactly, as the model uses fixed class indices. Thanks for sharing your working approach with the community! 🚀
